lib/lambda_ethereum_consensus/state_transition/state_transition.ex

```diff
@@ -3,14 +3,120 @@ defmodule LambdaEthereumConsensus.StateTransition do
   State transition logic.
   """
 
-  alias SszTypes.{BeaconState, SignedBeaconBlock}
+  alias LambdaEthereumConsensus.StateTransition
+  alias LambdaEthereumConsensus.StateTransition.EpochProcessing
+  alias SszTypes.{BeaconBlockHeader, BeaconState, SignedBeaconBlock}
 
+  @spec state_transition(BeaconState.t(), SignedBeaconBlock.t(), boolean()) ::
+          {:ok, BeaconState.t()} | {:error, String.t()}
   def state_transition(
         %BeaconState{} = state,
-        %SignedBeaconBlock{message: _block} = _signed_block,
+        %SignedBeaconBlock{message: block} = signed_block,
         _validate_result
       ) do
-    # TODO: implement
+    # NOTE: we aren't in a state to make validations yet
+    validate_result = false
+
+    state
+    # Process slots (including those with no blocks) since block
+    |> process_slots(block.slot)
+    # Verify signature
+    |> if_then_update(validate_result, fn state ->
+      if verify_block_signature(state, signed_block) do
+        {:ok, state}
+      else
+        {:error, "invalid block signature"}
+      end
+    end)
+    # Process block
+    |> map(&process_block(&1, block))
+    # Verify state root
+    |> map(
+      &if_then_update(&1, validate_result, fn state ->
+        if block.state_root != Ssz.hash_tree_root(state) do
+          {:error, "mismatched state roots"}
+        else
+          {:ok, state}
+        end
+      end)
+    )
+  end
+
+  defp process_slots(%BeaconState{slot: old_slot}, slot) when old_slot >= slot,
+    do: {:error, "slot is older than state"}
+
+  defp process_slots(%BeaconState{slot: old_slot} = state, slot) do
+    Enum.reduce((old_slot + 1)..slot, state, fn next_slot, state ->
+      state
+      |> process_slot()
+      # Process epoch on the start slot of the next epoch
+      |> if_then_update(rem(next_slot, ChainSpec.get("SLOTS_PER_EPOCH")) == 0, &process_epoch/1)
+      |> then(&%BeaconState{&1 | slot: next_slot})
+    end)
+  end
+
+  defp process_slot(%BeaconState{} = state) do
+    # Cache state root
+    previous_state_root = Ssz.hash_tree_root(state)
+    slots_per_historical_root = ChainSpec.get("SLOTS_PER_HISTORICAL_ROOT")
+    cache_index = rem(state.slot, slots_per_historical_root)
+    roots = List.replace_at(state.state_roots, cache_index, previous_state_root)
+    state = %BeaconState{state | state_roots: roots}
+
+    # Cache latest block header state root
+    state =
+      if state.latest_block_header.state_root == <<0::256>> do
```

Comment:
nit: eventually we should have the default ssz values as a function inside the ssz type, in this case `SszTypes.Root.default()`

Reply:
#395. I agree